rust/actyx/ax-core/src/runtime/operation/validation.rs

```diff
@@ -95,34 +95,68 @@ fn validate_string(value: &CborValue, string_refinement: &Option<String>) -> Res
 }
 
 /// Check if a CBOR value is an array. Can also be used to check for tuples (following RFC 7049).
-fn validate_array(value: &CborValue) -> Result<(), anyhow::Error> {
-    // NOTE(duarte): this validate is incomplete because we're not supporting subtyping, hence, we're just checking for arrays
-    if value.as_array().is_some() {
+fn validate_array(value: &CborValue, ty: &Type) -> anyhow::Result<()> {
+    if let Some(values) = value.as_array() {
+        for (i, value) in values.iter().enumerate() {
+            if let Err(err) = validate(&value.decode(), ty) {
+                return Err(anyhow::anyhow!("type mismatch, element {} is not of type {:?}", i, ty)).context(err);
+            }
+        }
         Ok(())
     } else {
         Err(anyhow::anyhow!("type mismatch, expected value to be an array"))
     }
 }
 
-fn validate_tuple(value: &CborValue, length: usize) -> Result<(), anyhow::Error> {
+fn validate_tuple(value: &CborValue, ty: &NonEmptyVec<Type>) -> anyhow::Result<()> {
```

Comment:
since the non-emptiness isn’t used in this function, it would be better to accept `&[Type]`